Update demos to make use of the new qml.sample and qml.expval(H) functionality (story #: 8118)

Jaybsoni opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [n] Ensure that your tutorial executes correctly, and conforms to the
      guidelines specified in the [README](../README.md).

- [y] Add a thumbnail link to your tutorial in `beginner.rst`, or if a
      QML implementation, in `implementations.rst`.

- [y] All QML tutorials conform to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      To auto format files, simply `pip install black`, and then
      run `black -l 100 path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Update demos to use qml.sample and qml.expval(H):**

**Summary:**
Updating the tutorials to make use of the new functionality in qml.sample and removing the dependance on qml.ExpvalCost as it will be redacted

**Relevant references:**
story id: 8118

**Possible Drawbacks:**

**Related GitHub Issues:**

Jaybsoni commented on demonstrations/tutorial_measurement_optimize.py:
```diff
@@ -141,14 +141,18 @@
 params = np.random.normal(0, np.pi, len(singles) + len(doubles))
-print("Cost function value:", cost(params))
+print("Cost function value:", cost_circuit(params))
 print("Number of quantum evaluations:", dev.num_executions)
```
Two things I would like to point out regarding the updates to this file: 
- This particular line of code (156), in the original tutorial produces 15 'executions'. The updated one only generates a single 'execution'. Not sure if this is a bug. If it isn't then I am not sure if we still want to keep this line of code. 
- Secondly, some of the results generated (in this entire tutorial) vary from the results of the original tutorial in the last 10^-13 ~ 10^-15 decimal places. I assume this is just noise on the python side. Not sure if that is worth looking into.

josh146 replied:
Good catches @Jaybsoni! I can provide some insights.

> This particular line of code (156), in the original tutorial produces 15 'executions'. The updated one only generates a single 'execution'. Not sure if this is a bug. If it isn't then I am not sure if we still want to keep this line of code.

This is because `return expval(H)` acts very differently to `ExpvalCost` internally. While `ExpvalCost` always creates a separate circuit per Hamiltonian commuting group (as you need to do on hardware), `return expval(H)` tries to be smarter:

- If `shots=None`, it computes the expectation value `<psi|H|psi>` _directly_, without needing to group Hamiltonian terms.
- If `shots>0`, it reverts to the existing behaviour of grouping.

In most case when running a simulation, we only care about the expectation value --- _how_ we computed the expectation value is not so important. But in this case, this particular demo actually _explores_ the grouping behaviour!

So we want to preserve the old behaviour here. We have several options:

1. One option is to _manually_ group the Hamiltonian, and print the commuting groups:

  ```pycon
  >>> H.compute_grouping()
  >>> len(H.grouping_indices)
  5
  ```

  (and ignore that the simulation is actually just doing one execution!). I don't really like this though, doesn't feel very clear.

2. Another option is to set `shots=1000` (or something similar) for the sections that use `ExpvalCost` (e.g., the first section, and the very last code-block). The rest can be left at `shots=None` for speed/accuracy.

3. A third option, which I think I like the most (but which doesn't exist yet!) is to extend PennyLane so that you can specify that you want grouping to occur even if `shots=None`. But this requires (a) thinking about how we want the UI to look, and (b) changes to PL master 🙂 

> Secondly, some of the results generated (in this entire tutorial) vary from the results of the original tutorial in the last 10^-13 ~ 10^-15 decimal places. I assume this is just noise on the python side. Not sure if that is worth looking into.

Yep, this would be related to the different logical pathway taken by `expval(H)` and `ExpvalCost` 🙂

josh146 replied:
Another tangential thought: I just realized that `dev.num_executions` is now outdated, and has been replaced with the device tracker: https://pennylane.readthedocs.io/en/stable/code/api/pennylane.Tracker.html

Jaybsoni replied:
Cool! I replaced `dev.num_executions` with `qml.Tracker`. I went with your 2nd suggestion as the 3rd one seems to be worth a separate PR. 

Also wanted to mention that I didn't set the shots for the very last code block because I don't think it is required, I think it explicitly computes the groupings.

Jaybsoni commented on demonstrations/tutorial_vqe_qng.py:
```diff
@@ -63,7 +63,11 @@ def circuit(params, wires=0):
 H = qml.Hamiltonian(coeffs, obs)
-cost_fn = qml.ExpvalCost(circuit, H, dev)
+
+@qml.qnode(dev)
```
The documentation paragraph associated with this bit of code (lines 58 - 60) describes some capabilities of qml.ExpvalCost, not sure if this is still true for qml.expval. Let me know how I should update that bit of documentation.

josh146 replied:
good catch @Jaybsoni, I would remove the "using the ``ExpvalCost`` class," part :)

Jaybsoni replied:
Ah awesome thanks !

Jaybsoni commented on demonstrations/tutorial_vqe_parallel.py:
```diff
@@ -178,10 +187,9 @@ def circuit(param, wires):
+energies = [get_expval_qnode(h) for h in hamiltonians]
 
 ##############################################################################
 # Calculating the potential energy surface
```
The paragraph below also describes some functionality which is specific to ExpvalCost, any suggestions on how we can edit this one?

antalszava replied:
Oh, that's a good question :thinking: Frankly, I'm not sure if using `parallel` option of QNodeCollection is still accessible with the `expval(H)` syntax.

@josh146 any thoughts?

josh146 replied:
I don't think it is! Perhaps best to leave this demo using `ExpvalCost`?

Jaybsoni replied:
Cool, I will revert the changes for this demo then!

Jaybsoni commented on demonstrations/learning2learn.py:
```diff
@@ -239,8 +240,9 @@ def hamiltonian(params, **kwargs):
-        # ExpvalCost evaluates the expectation value of an operator
-        cost = qml.ExpvalCost(circuit, cost_h, dev, interface="tf", diff_method="backprop")
+        # This qnode evaluates the expectation value of the cost hamiltonian operator
+        # cost = qml.ExpvalCost(circuit, cost_h, dev, interface="tf", diff_method="backprop")
```
It seems ExpvalCost has keyword arguments which (I assume) specify how derivatives are computed of the underlying expectation value with respect to the trainable parameters. How can I preserve these settings otherwise ?

antalszava replied:
It should be possible to simply pass them to the QNode:

```python

@qml.qnode(dev, interface="tf", diff_method="backprop")
def circuit():
    # some circuit
    return expval(H)
```

Jaybsoni replied:
Ahh I see, thanks !

antalszava commented on demonstrations/tutorial_measurement_optimize.py:
```diff
@@ -141,15 +141,20 @@
 params = np.random.normal(0, np.pi, len(singles) + len(doubles))
-print("Cost function value:", cost(params))
-print("Number of quantum evaluations:", dev.num_executions)
+with qml.Tracker(dev) as tracker:  # track the number of executions
+    print("Cost function value:", cost_circuit(params))
+
+print("Number of quantum evaluations:", tracker.totals['executions'])
```
Nice!

antalszava commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -176,15 +176,13 @@ def U_C(gamma):
-# a measurement of multiple qubits in the computational basis, so we define
-# a Hermitian operator to do this. The eigenvalues of the operator are
-# the qubit measurement values in integer form.
+# We will need a way to take our sampled measurement, which is a bitstring
+# representing the measured state in the computational basis, and convert it
+# to an int
```
```suggestion
# We will need a way to sample a measurement of multiple qubits in the computational basis
#  in integer form.
```

antalszava reviewed (commented):
Hi @Jaybsoni, this is looking nice! :tada: :slightly_smiling_face: Left some questions/suggestions.

1. There seem to be some occurrences of `ExpvalCost` in https://github.com/PennyLaneAI/qml/blob/master/demonstrations/tutorial_vqe_parallel.py too. Would they need to be changed?
2. Did the changed demos produce the expected outputs? (Including the ones that are not automatically run.)

Jaybsoni commented on demonstrations/tutorial_vqls.py:
```diff
@@ -474,20 +474,23 @@ def prepare_and_sample(weights):
+for sam in raw_samples:
+    samples.append(int("".join(str(bs) for bs in sam), base=2))
 
+q_probs = np.bincount(samples, minlength=2 ** n_qubits) / n_shots
```
I think I specified it because I wanted to cover the case where we don't sample some of the bit strings at either end of the distributions. If we miss it, then the distribution might have fewer bins. But I guess its not necessary in this case as it wasn't included last time?

antalszava replied:
Yep, maybe best left as is, if the change is not required. :)

Jaybsoni replied:
Got it !

Jaybsoni commented on demonstrations/tutorial_measurement_optimize.py:
```diff
@@ -125,7 +125,7 @@
 
 # Create a 4 qubit simulator
-dev = qml.device("default.qubit", wires=num_qubits)
+dev = qml.device("default.qubit", wires=num_qubits, shots=1000)
```
Because when shots=None, the result is computed analytically the tracker only registers a single execution, but in this tutorial we want to track the number of executions if we were to explicitly run it.

Jaybsoni commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -315,15 +315,18 @@ def circuit(params, **kwargs):
 #
 # The cost function is the expectation value of :math:`H_C`, which we want to minimize. The
-# function :func:`~.pennylane.ExpvalCost` is designed for this purpose: it returns the
+# function :func:`~.pennylane.expval(H)` is designed for this purpose: it returns the
```
Got it,

Jaybsoni commented on demonstrations/tutorial_measurement_optimize.py:
```diff
@@ -745,12 +750,15 @@ def circuit(weights, group=None, **kwargs):
 H = qml.Hamiltonian(coeffs=np.ones(len(terms)), observables=terms)
-cost_fn = qml.ExpvalCost(qml.templates.StronglyEntanglingLayers, H, dev, optimize=True)
+@qml.qnode(dev)
+def cost_fn(params):
+    qml.templates.StronglyEntanglingLayers(params, wires=dev.wires)
+    return qml.expval(H)
```
Ahh I see!

Jaybsoni commented on demonstrations/tutorial_falqon.py:
```diff
@@ -234,6 +236,9 @@ def ansatz(beta, **kwargs):
             delta_t=delta_t
         )
 
+        if measurement_h is not None:
+            return qml.expval(measurement_h)
+
```
Got it!

Jaybsoni commented on demonstrations/tutorial_falqon.py:
```diff
@@ -410,14 +411,14 @@ def qaoa_layer(gamma, beta):
 
-# Creates the full QAOA circuit
+# Creates the full QAOA circuit as an executable cost function
+@qml.qnode(dev)
```
Ok, I think I get what you are suggesting, I implemented it in, let me know if this is what you had in mind.

Jaybsoni commented:
> Hi @Jaybsoni, this is looking nice! 🎉 🙂 Left some questions/suggestions.
> 
> 1. There seem to be some occurrences of `ExpvalCost` in https://github.com/PennyLaneAI/qml/blob/master/demonstrations/tutorial_vqe_parallel.py too. Would they need to be changed?
> 2. Did the changed demos produce the expected outputs? (Including the ones that are not automatically run.)

Hey Antal, I believe we discussed with Josh that we should leave the vqe_parallel tutorial as is. I have addressed all of your comments above. I confirmed that all of the automatically running demos are working as expected. I will go over the rest of them today and get back to you.

josh146 commented:
> I believe we discussed with Josh that we should leave the vqe_parallel tutorial as is.

Yep, the reason was that the parallel VQE tutorial is designed to highlight the Dask parallelization of `ExpvalCost`, which no longer exists in the `expval(H)` pipeline, we may have to revisit this demo later

Jaybsoni commented:
Hi @antalszava and @josh146,

I finished checking most of the demos, there are just 2 left to check which I am having trouble running. 

1st: [qsim beyond classical demo](https://pennylane.ai/qml/demos/qsim_beyond_classical.html)
In this case I am running into a weird import error with qsimcirq: 
<img width="581" alt="image" src="https://user-images.githubusercontent.com/21964864/134738391-32d6bddf-0043-4918-be7e-60068d34f51e.png">
<img width="577" alt="image" src="https://user-images.githubusercontent.com/21964864/134738543-61edc511-9333-4dad-9694-3d4cc543ac78.png">
<img width="574" alt="image" src="https://user-images.githubusercontent.com/21964864/134738613-4a4c0627-8c0d-4852-a113-51ab9e241616.png">

Have either of you come across this error before ? 

2nd: [braket parallel gradients](https://pennylane.ai/qml/demos/braket-parallel-gradients.html)
This one requires a Braket account to complete. 

If you could help me debug the first issue and provide me with at Braket account I can finish the reviews my self, otherwise it might be faster if someone who has a working setup can review them quickly so we can get this merged in? I am 90% sure it shouldn't cause any issues. 

Cheers,

josh146 commented:
>  In this case I am running into a weird import error with qsimcirq:

I'm not 100% sure here, but I think @rmoyard or @thisac might know, I think they have run into similar issues!

>  This one requires a Braket account to complete.

Sure, that can be arranged!

mariaschuld commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -176,15 +176,12 @@ def U_C(gamma):
-# a Hermitian operator to do this. The eigenvalues of the operator are
-# the qubit measurement values in integer form.
+# We will need a way to sample a measurement of multiple qubits in the computational basis
+#  in integer form.
```
```suggestion
# in integer form.
```

mariaschuld commented on demonstrations/tutorial_vqe_qng.py:
```diff
@@ -291,9 +295,12 @@ def ansatz(params, wires=[0, 1, 2, 3]):
 # Note that the qubit register has been initialized to :math:`|1100\rangle`, which encodes for
 # the Hartree-Fock state of the hydrogen molecule described in the minimal basis.
-# Again, we define the cost function using the ``ExpvalCost`` class.
+# Again, we define the cost function using ``expval(H)``.
```
I was wondering if here and in other places it would be clearer to say "we define the cost to be the expectation of the Hamiltonian" or "we define the cost to be the QNode itself"...It's not so important though.

Jaybsoni replied:
That feels clearer to me as well,

mariaschuld commented on demonstrations/tutorial_vqe_spin_sectors.py:
```diff
@@ -185,17 +185,23 @@ def circuit(params, wires):
 
 ##############################################################################
-# Next, we use the :class:`~.pennylane.ExpvalCost` class to define the cost function.
+# Next, we use the :func:`~.pennylane.expval` function to define the cost function.
```
For example, we are not using the expval function, but the entire qnode to define the cost. Or in other words, the measurement outcome.

Jaybsoni replied:
Agreed!

mariaschuld reviewed (commented):
Thanks for taking on this cumbersome task @Jaybsoni! 

@josh146  and I had a chat about how to deal with demos that use `ExpvalCost(..,parallel=True)` and we were wondering if one should just "unpack" the ExpvalCost and use dask explicitly. I.e., one could make one Qnode per Hamiltonian term, and then execute all Qnodes with dask. What do you think? 

This could be useful for users to get used to dask, and should have the same performance! It also shows an alternative, manual way to do VQE...

If you agree, just ping me if there are any problems! The important dask code is in the `QNodeCollection` class if I remember correctly.

antalszava commented:
@Jaybsoni just a guess, could there be something going on with the [`libomp` installation](https://formulae.brew.sh/formula/libomp)? That image seems to be missing. :thinking:

If it doesn't come around, let us know. :slightly_smiling_face:

thisac commented:
>  In this case I am running into a weird import error with qsimcirq:

I don't recognize the error, so I'm not sure what's causing it. Have you tried both installing with `pip` as well as directly from the GitHub master branch? I think I remember only getting one of them to work before.

@Jaybsoni I'm able to run the demo fine on my personal laptop, so I could help testing it out if you'd like.

Jaybsoni commented:
Hey all (@antalszava, @josh146), 

I can now confirm that all of the demos are running as expected. I was able to run the Braket parallel gradients demo today and @antalszava ran the Qsim beyond classical demo. Beyond that I have addressed any comments left on my changes. 


> Thanks for taking on this cumbersome task @Jaybsoni!
> 
> @josh146 and I had a chat about how to deal with demos that use `ExpvalCost(..,parallel=True)` and we were wondering if one should just "unpack" the ExpvalCost and use dask explicitly. I.e., one could make one Qnode per Hamiltonian term, and then execute all Qnodes with dask. What do you think?

@mariaschuld, thank you for the comments! That does seem to be a more informative solution, I think it might be better suited as its own separate story/ticket which I can take on after merging this one. 

I think the everything is good to go, let me know if you have any other suggestions and I can address them. 
Cheers,

antalszava commented:
[ch8118]

antalszava commented on demonstrations/braket-parallel-gradients.py:
```diff
@@ -96,7 +96,7 @@
 """
 
-my_bucket = "amazon-braket-Your-Bucket-Name"  # the name of the bucket
+my_bucket = "amazon-braket-Your-Bucket-Name"  # the name of the bucket, keep the 'amazon-braket-' and then include name
```
```suggestion
my_bucket = "amazon-braket-Your-Bucket-Name"  # the name of the bucket, keep the 'amazon-braket-' prefix and then include the bucket name
```

antalszava commented on demonstrations/tutorial_falqon.py:
```diff
@@ -410,14 +411,17 @@ def qaoa_layer(gamma, beta):
+# Creates the full QAOA circuit as an executable cost function
 def qaoa_circuit(params, **kwargs):
-    for w in dev.wires:
+    for w in new_graph.nodes:
```
How come this was changed?

Jaybsoni replied:
Ah nice catch @antalszava, it must be a residual change left over when I was experimenting with the demo. I was sure I changed it back but I guess I must have missed it.

antalszava commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -219,12 +220,14 @@ def circuit(params, wires):
-    return qml.ExpvalCost(circuit, H(x), dev)(params)
+    hamiltonian = H(x)
+    run_circuit = qml.QNode(circuit, dev)
+    return run_circuit(params, obs=hamiltonian, wires=range(6))
```
It might be worth creating an object for the number of wires e.g., on line 219 (`num_wires = 6`). Then this object can be used in the device definition and when passing the wires to `circuit`.

This would help with:
* Allowing that the number of wires needs to be changed only at one place (if applicable);
* Making the code slightly more descriptive by using a name such as `num_wires` or so.

Jaybsoni replied:
Ahh yes, it improves code readability!

antalszava reviewed (approved):
Looks good to me :champagne: :tada: :blush: Awesome addition @Jaybsoni, thank you! No major comments remain from my side.

Just double-checking: might it be that we're restoring the old cached outputs for the automatically executed demos? (ccing @josh146)

Jaybsoni commented:
> Looks good to me 🍾 🎉 😊 Awesome addition @Jaybsoni, thank you! No major comments remain from my side.
> 
> Just double-checking: might it be that we're restoring the old cached outputs for the automatically executed demos? (ccing @josh146)

Thanks for the comments! I made sure to run each demo script locally as opposed to relying on the CI tests, so I can confidently say that each of the demos that have been changed here should run (at least locally on my setup). Let me know if there is anything else.

mariaschuld commented on demonstrations/tutorial_coherent_vqls.py:
```diff
@@ -496,18 +496,22 @@ def prepare_and_sample(weights):
-    basis_obs = qml.Hermitian(np.diag(range(2 ** n_qubits)), wires=range(n_qubits))
-
-    return qml.sample(basis_obs)
+    # Therefore, sampling the device will give us a value of 0 or 1 for each qubit (n_qubits)
```
```suggestion
    # Therefore, sampling from the device will give us a value of 0 or 1 for each qubit (n_qubits)
```

I think this is more correct :)

mariaschuld commented on demonstrations/qsim_beyond_classical.py:
```diff
@@ -297,8 +297,8 @@ def generate_single_qubit_gate_list():
 # argument to our circuit allowing us to switch between the two returns. We
-# sample from the Pauli-Z observable on all wires, which will give us the
-# eigenvalues :math:`\pm 1` of the observable, corresponding to the states
+# take samples from the computational basis from all wires, which will give us
```
```suggestion
# take samples from the computational basis state using all wires, which will return bitstrings of
```

mariaschuld commented on demonstrations/tutorial_coherent_vqls.py:
```diff
@@ -496,18 +496,22 @@ def prepare_and_sample(weights):
-samples = prepare_and_sample(w).astype(int)
+raw_samples = prepare_and_sample(w)
+
+# Here we convert the raw samples (bit strings) into integers and count them
```
```suggestion
# convert the raw samples (bit strings) into integers and count them
```

Not so important... :)

mariaschuld commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -216,15 +217,18 @@ def circuit(params, wires):
-# Hamiltonian parameters.
+# Next, we define the `cost`` function :math:`g(\theta, x)` which depends on
+# both the circuit and the Hamiltonian parameters. Specifically we consider the
+# expectation values of the hamiltonian.
```
Nice!

josh146 replied:
```suggestion
# expectation values of the Hamiltonian.
```

mariaschuld commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -314,16 +314,19 @@ def circuit(params, **kwargs):
-# performed. We use the PennyLane-Qulacs plugin to
-# run the circuit on the Qulacs simulator:
+# The cost function is the expectation value of :math:`H_C`, which we want to minimize. We
+# use the function :func:`~.pennylane.expval` by passing it a Hamiltonian `H`: it returns the
```
```suggestion
# use the function :func:`~.pennylane.expval` which returns the
```

I am not sure if "passing it" is grammatically correct? Could just be simplified like this...

mariaschuld commented on demonstrations/tutorial_vqe_qng.py:
```diff
@@ -291,9 +295,12 @@ def ansatz(params, wires=[0, 1, 2, 3]):
 # Note that the qubit register has been initialized to :math:`|1100\rangle`, which encodes for
 # the Hartree-Fock state of the hydrogen molecule described in the minimal basis.
-# Again, we define the cost function using the ``ExpvalCost`` class.
+# Again, we define the cost function to be the following QNode using the ``expval(H)`` function.
```
```suggestion
# Again, we define the cost function to be the following QNode that measures ``expval(H)``:
```

Only a suggestion...

mariaschuld commented on demonstrations/tutorial_vqls.py:
```diff
@@ -474,20 +474,23 @@ def prepare_and_sample(weights):
-q_probs = np.bincount(samples) / n_shots
+raw_samples = prepare_and_sample(w)
 
+# Here we convert the raw samples (bit strings) into integers and count them
```
```suggestion
# convert the raw samples (bit strings) into integers and count them
```
just a suggestion

mariaschuld reviewed (approved):
Again, thanks so much for taking this on! I did not check if the demos return the correct results, and I added some suggestions to polish the comments, but they are not blocking. Other than that it looks good to me!

josh146 commented on demonstrations/tutorial_chemical_reactions.py:
```diff
@@ -297,7 +301,7 @@ def circuit(params, wires):
+        params, energy = opt.step_and_cost(circuit, params)
         if np.abs(energy - prev_energy) < 1e-6:
             break
         prev_energy = energy
```
By switching to `expval(H)`, this demo went from 6 minutes to 1 minute execution time!

![image](https://user-images.githubusercontent.com/2959003/135223142-e842fdd1-455a-4009-99fa-e8165883ee42.png)

![image](https://user-images.githubusercontent.com/2959003/135223153-81fb926a-51da-4948-9219-1d49b5eff94d.png)

Jaybsoni replied:
Thats amazing!

josh146 commented on demonstrations/tutorial_falqon.py:
```diff
@@ -447,11 +451,11 @@ def qaoa_circuit(params, **kwargs):
+probs = prob_circuit(params)
 plt.bar(range(2**len(dev.wires)), probs)
 plt.xlabel("Bit string")
 plt.ylabel("Measurement Probability")
```
This one is another amazing speed improvement 😍 

![image](https://user-images.githubusercontent.com/2959003/135223580-e445c5f9-be4d-4f8e-b7a8-173f7657aa7c.png)

![image](https://user-images.githubusercontent.com/2959003/135223595-f6d685e9-f2ad-43aa-afd5-c6ea7dcb380e.png)

josh146 replied:
@Jaybsoni oddly enough, after your changes here, the output is also _better_; the state |112> has a higher probability!

Jaybsoni replied:
Interesting!

josh146 commented on demonstrations/tutorial_measurement_optimize.py:
```diff
@@ -141,15 +141,20 @@
+print("Number of quantum evaluations:", tracker.totals['executions'])
 
 ##############################################################################
 # How about a larger molecule? Let's try the water molecule :download:`h2o.xyz </demonstrations/h2o.xyz>`:
```
@Jaybsoni is there a reason for leaving this at the bottom?

![image](https://user-images.githubusercontent.com/2959003/135224203-4f0af70f-7993-41ae-aacb-b8ca74c881af.png)

@mariaschuld, I suppose in order to change this, we need to modify `qml.expval(H)` in order to optionally support grouping when requested with `shots=None`?

Jaybsoni replied:
I think we decide to leave it in because the `optimize=True` keyword argument wasn't accessible through `expval(H)`? 

> Likely we'd have to keep this as is as if I recall correctly `optimize=False` is the default under the hood when using a QNode & `qml.expval(H)`.

josh146 replied:
Thanks for the reminder!

josh146 commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -188,11 +188,12 @@ def H(x):
 # The ``hf`` array is used by the :class:`~.pennylane.BasisState` operation to initialize
```
![image](https://user-images.githubusercontent.com/2959003/135224411-10e9bf46-264c-4eb3-ba63-5c52f46f4a1a.png)

![image](https://user-images.githubusercontent.com/2959003/135224429-30b78068-eedf-4d22-9473-9e54828d5fb1.png)

😻

josh146 commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -216,15 +217,18 @@ def circuit(params, wires):
-# Next, we use the PennyLane class :class:`~.pennylane.ExpvalCost` to define the
-# ``cost`` function :math:`g(\theta, x)` which depends on both the circuit and the
-# Hamiltonian parameters.
+# Next, we define the `cost`` function :math:`g(\theta, x)` which depends on
```
```suggestion
# Next, we define the ``cost`` function :math:`g(\theta, x)` which depends on
```

josh146 commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -216,15 +217,18 @@ def circuit(params, wires):
+# expectation values of the hamiltonian.
 
 def cost(params, x):
-    return qml.ExpvalCost(circuit, H(x), dev)(params)
+    hamiltonian = H(x)
+    run_circuit = qml.QNode(circuit, dev)
+    return run_circuit(params, obs=hamiltonian, wires=range(num_wires))
```
@Jaybsoni since it appears that the QNode is only really needed to be created once, should we instead use the decorator above,

```python
num_wires = 6
dev = qml.device("default.qubit", wires=num_wires)

@qml.qnode(dev)
def circuit(params, obs, wires):
    qml.BasisState(hf, wires=wires)
    qml.DoubleExcitation(params[0], wires=[0, 1, 2, 3])
    qml.DoubleExcitation(params[1], wires=[0, 1, 4, 5])
    return qml.expval(obs)
```
and then here, we can simply have

```python
def cost(params, x):
    hamiltonian = H(x)
    return circuit(params, obs=hamiltonian, wires=range(num_wires)
```

Jaybsoni replied:
Ahh yes this is much cleaner. I didn't think of moving the device definition up to the circuit definition because in a few of the demos the device was example specific, but in this case its just a standard 6 wire device.

josh146 commented on demonstrations/tutorial_mol_geo_opt.py:
```diff
@@ -240,14 +244,13 @@ def cost(params, x):
 def grad_x(x, params):
     grad_h = qml.finite_diff(H)(x)
-    grad = [qml.ExpvalCost(circuit, obs, dev)(params) for obs in grad_h]
+    grad = [qml.QNode(circuit, dev)(params, obs=obs, wires=range(num_wires)) for obs in grad_h]
```
The above suggestion would also make this line more efficient:

```python
grad = [circuit(params, obs=obs, wires=range(num_wires)) for obs in grad_h]
```

Jaybsoni replied:
Got it

josh146 commented on demonstrations/tutorial_qaoa_intro.py:
```diff
@@ -314,16 +314,19 @@ def circuit(params, **kwargs):
 # ``params[0]`` and ``params[1]`` into each layer of the circuit. That's it! The last
```
@mariaschuld this demo is actually a good example of the slowdown incurred for QAOA using the new sparse default when `shots=None`:

Before:
![image](https://user-images.githubusercontent.com/2959003/135225127-159a7599-efb3-4819-b8fb-232a01c007e5.png)

After:
![image](https://user-images.githubusercontent.com/2959003/135225147-b79b2949-b42d-4509-8566-683049e5c241.png)

Makes me think we should definitely consider how to avoid the sparse pipeline when all terms commute

josh146 commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -176,15 +176,12 @@ def U_C(gamma):
 
```
Interestingly, since we are using `shots=1` here, this demo has no speed increase/decrease :)

josh146 commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -176,15 +176,12 @@ def U_C(gamma):
-
-def comp_basis_measurement(wires):
-    n_wires = len(wires)
-    return qml.Hermitian(np.diag(range(2 ** n_wires)), wires=wires)
```
So glad this is finally gone!

josh146 commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -176,15 +176,12 @@ def U_C(gamma):
-# a measurement of multiple qubits in the computational basis, so we define
-# a Hermitian operator to do this. The eigenvalues of the operator are
-# the qubit measurement values in integer form.
+# We will need a way to sample a measurement of multiple qubits in the computational basis
+# in integer form.
```
```suggestion
# We will need a way to convert a bitstring, representing a sample of multiple qubits
# in the computational basis, to integer or base-10 form.
```
Perhaps this is slightly clearer?

josh146 commented on demonstrations/tutorial_qaoa_maxcut.py:
```diff
@@ -218,7 +215,7 @@ def circuit(gammas, betas, edge=None, n_layers=1):
     if edge is None:
         # measurement phase
-        return qml.sample(comp_basis_measurement(range(n_wires)))
+        return qml.sample()
```
🔥

josh146 commented on demonstrations/tutorial_vqe_qng.py:
```diff
@@ -55,15 +55,19 @@ def circuit(params, wires=0):
 
```
![image](https://user-images.githubusercontent.com/2959003/135226012-b3922d5b-8b1d-4cf6-9dce-2fc4c98ba981.png)

![image](https://user-images.githubusercontent.com/2959003/135226031-cb6464c0-16ff-4539-97b2-1f301688233d.png)

💪

josh146 reviewed (changes requested):
Awesome work @Jaybsoni! I've left some very minor suggestions, once these have been considered, happy to approve 💪

antalszava reviewed (approved):
:100:

josh146 reviewed (approved):
Looks great @Jaybsoni 💪
